runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java

```diff
@@ -247,25 +247,27 @@ public void reduce(Iterable<WindowedValue<InputT>> iterable, Collector<RawUnionV
     timerInternals.advanceSynchronizedProcessingTime(BoundedWindow.TIMESTAMP_MAX_VALUE);
 
     // Now we fire the timers and process elements generated by timers (which may be timers itself)
-    try (RemoteBundle bundle =
-        stageBundleFactory.getBundle(
-            receiverFactory, timerReceiverFactory, stateRequestHandler, progressHandler)) {
-
-      PipelineTranslatorUtils.fireEligibleTimers(
-          timerInternals,
-          (KV<String, String> transformAndTimerId, Timer<?> timerValue) -> {
-            FnDataReceiver<Timer> fnTimerReceiver =
-                bundle.getTimerReceivers().get(transformAndTimerId);
-            Preconditions.checkNotNull(
-                fnTimerReceiver, "No FnDataReceiver found for %s", transformAndTimerId);
-            try {
-              fnTimerReceiver.accept(timerValue);
-            } catch (Exception e) {
-              throw new RuntimeException(
-                  String.format(Locale.ENGLISH, "Failed to process timer: %s", timerValue));
-            }
-          },
-          currentTimerKey);
+    while (timerInternals.hasPendingTimers()) {
```

Comment:
It makes sense to do this in batch as well since the processing timers should get dropped and any newly scheduled event time timers should be continually fired.

This would require updating FlinkExecutableStageFunction and SparkExecutableStageFunction

Reply:
Also, I made this same comment earlier today and it was lost somehow.

Reply:
Agree that this should apply to Spark runner too. To clarify, FlinkExecutableStageFunction is for batch only, meaning this change only affected Flink batch. I assume Flink streaming (ExecutableStageDoFnOperator) was working from the start.

Reply:
Made #11595 to update Spark.

Reply:
Yes, this is a batch-only change (also evident in the commit message). The streaming operator already does that. 

> processing timers should get dropped and any newly scheduled event time timers should be continually fired.

We do not drop processing timers. We fire all timers, including new ones, until there are none left.